### ISSUE TITLE
Deterministic instantiation of endpoint_a and endpoint_b in Link

### ISSUE THREAD

**Issue by italovalcy** [kytos-ng/kytos#221]
Currently, the `Link.id` property utilizes a deterministic approach to store and calculate the link ID, based on [ascending sort of the DPIDs and ports](https://github.com/kytos-ng/kytos/blob/f05059fd52fd5ec64e3bfe3ce62bf87cb6e50ec3/kytos/core/link.py#L75-L86).

On the other hand, the attributes `Link.endpoint_a` and `Link.endpoint_b` are not sorted. This leads to situations where, depending on which event was handled first by of_lldp, the same link can be stored in different order in the endpoints:

```
## start kytos and do some updates in the topology...

root@7767801619fb:/# curl -s http://127.0.0.1:8181/api/kytos/topology/v3/ > topology1.json

## eventually restart kytos and wait topology to be loaded

root@7767801619fb:/# curl -s http://127.0.0.1:8181/api/kytos/topology/v3/ > topology2.json

root@7767801619fb:/# diff -Naur <(jq --sort-keys . topology1.json) <(jq --sort-keys . topology2.json)
--- /dev/fd/63	2022-05-20 20:34:16.829761741 +0000
+++ /dev/fd/62	2022-05-20 20:34:16.829761741 +0000
@@ -7,38 +7,38 @@
         "endpoint_a": {
           "active": true,
           "enabled": true,
-          "id": "00:00:00:00:00:00:00:18:11",
+          "id": "00:00:00:00:00:00:00:11:11",
           "link": "1e7e531ae50f419b1d35d311a66351d997ea567348b691d8dccdc639e1f8340a",
           "lldp": true,
-          "mac": "aa:72:80:89:7a:af",
+          "mac": "26:6e:bc:27:57:5b",
           "metadata": {
             "mtu": 9000,
-            "port_name": "Interface-18:11"
+            "port_name": "Interface-11:11"
           },
-          "name": "Ampath4-eth11",
+          "name": "Ampath1-eth11",
           "nni": true,
           "port_number": 11,
           "speed": 1250000000,
-          "switch": "00:00:00:00:00:00:00:18",
+          "switch": "00:00:00:00:00:00:00:11",
           "type": "interface",
           "uni": false
         },
         "endpoint_b": {
           "active": true,
           "enabled": true,
-          "id": "00:00:00:00:00:00:00:11:11",
+          "id": "00:00:00:00:00:00:00:18:11",
           "link": "1e7e531ae50f419b1d35d311a66351d997ea567348b691d8dccdc639e1f8340a",
           "lldp": true,
-          "mac": "26:6e:bc:27:57:5b",
+          "mac": "aa:72:80:89:7a:af",
           "metadata": {
             "mtu": 9000,
-            "port_name": "Interface-11:11"
+            "port_name": "Interface-18:11"
           },
-          "name": "Ampath1-eth11",
+          "name": "Ampath4-eth11",
           "nni": true,
           "port_number": 11,
           "speed": 1250000000,
-          "switch": "00:00:00:00:00:00:00:11",
+          "switch": "00:00:00:00:00:00:00:18",
           "type": "interface",
           "uni": false
         },
@@ -46,6 +46,7 @@
         "metadata": {
           "availability": 100,
           "bandwidth": 100,
+          "last_status_change": 1653078485.537819,
           "last_status_is_active": true,
           "latency": 1,
           "link_name": "Ampath4-eth11--Ampath1-eth11",
@@ -59,38 +60,38 @@
         "endpoint_a": {
           "active": true,
           "enabled": true,
-          "id": "00:00:00:00:00:00:00:15:6",
+          "id": "00:00:00:00:00:00:00:16:6",
           "link": "21c046f5eb9fbf577701974e207c2fd2ccd8cc4b91fae77ad396924b79c48483",
           "lldp": true,
-          "mac": "72:a1:47:7f:b4:15",
+          "mac": "76:18:c6:24:64:06",
           "metadata": {
             "mtu": 9000,
-            "port_name": "Interface-15:6"
+            "port_name": "Interface-16:6"
           },
-          "name": "AL2-eth6",
+          "name": "AL3-eth6",
           "nni": true,
           "port_number": 6,
           "speed": 1250000000,
-          "switch": "00:00:00:00:00:00:00:15",
+          "switch": "00:00:00:00:00:00:00:16",
           "type": "interface",
           "uni": false
         },
         "endpoint_b": {
           "active": true,
           "enabled": true,
-          "id": "00:00:00:00:00:00:00:16:6",
+          "id": "00:00:00:00:00:00:00:15:6",
           "link": "21c046f5eb9fbf577701974e207c2fd2ccd8cc4b91fae77ad396924b79c48483",
           "lldp": true,
-          "mac": "76:18:c6:24:64:06",
+          "mac": "72:a1:47:7f:b4:15",
           "metadata": {
             "mtu": 9000,
-            "port_name": "Interface-16:6"
+            "port_name": "Interface-15:6"
           },
-          "name": "AL3-eth6",
+          "name": "AL2-eth6",
           "nni": true,
           "port_number": 6,
           "speed": 1250000000,
-          "switch": "00:00:00:00:00:00:00:16",
+          "switch": "00:00:00:00:00:00:00:15",
           "type": "interface",
           "uni": false
         },
@@ -98,6 +99,7 @@
         "metadata": {
           "availability": 100,
           "bandwidth": 100,
+          "last_status_change": 1653078485.5223212,
           "last_status_is_active": true,
           "latency": 1,
           "link_name": "AL2-eth6--AL3-eth6",
...
```

Maybe we could sort the endpoint when instantiating a Link, instead of at the time of `Link.id` calculation.

Since this has no impact other than facilitating visualization and convenience, the criticality of this issue is very low.

**Comment by viniarck** [comment]
@Ktmi could you add this issue here on your radar? Since you've had extensive experience with Link IDs and related core parts, that should be easy for you to tackle. Much appreciated. 

It'd be great to have this one fixed for upcoming `2023.1` release.

**Comment by Ktmi** [comment]
I've been considering something related to this for a while. What could be done is to decouple the link from its endpoints, and instead have a reference to an object representing the overall topology. Accessing the endpoints can then be done by looking at LinkID, then accessing the interfaces through the topology. LinkID just needs the interfaces sorted ahead of the time to enforce determinism, which can be done with a few simple changes.

What I am proposing should approximately look like this:

```python

from typing_extensions import NewType
from dataclasses import dataclass
import hashlib

SwitchID = NewType('SwitchID', str)

@dataclass
class Switch:
    controller: 'Controller'
    id: SwitchID

    def interfaces(self):
        return self.controller.interfaces[self.id]

class InterfaceID(str):
    """Interface Identifier"""

    __slots__ = ("switch", "port")

    def __new__(cls, switch: SwitchID, port: int):
        return super().__new__(cls, f"{switch}:{port}")

    def __init__(self, switch: SwitchID, port: int):
        # Used for sorting, but can be accessed
        self.switch = switch
        self.port = port
        self.tuple = (switch, port)
        super().__init__()

    def __lt__(self, other):
        # Ensures that instances are sortable in a way that maintains backwards
        # compatibility when creating instances of LinkID
        if isinstance(other, InterfaceID):
            return self.tuple < other.tuple
        return NotImplemented

    def __getnewargs__(self):
        """To make sure it's pickleable"""
        return self.tuple

@dataclass
class Interface:
    controller: 'Controller'
    id: InterfaceID

    @property
    def switch(self):
        return self.controller.switches[self.id.switch]

    @property
    def link(self):
        # Yes this is very inefficent, but the alternative is to write
        # a really complicated specialized dict type.
        links = [link for link in self.controller.links if self.id in link.interfaces]
        if links:
            return self.controller.links[links[0]]
        return None

    @property
    def endpoint(self):
        link = self.link
        if link is None:
            return None
        endpoints = link.interfaces
        if link.endpoint_a.id == self.id:
            return link.endpoint_b
        return link.endpoint_a


class LinkID(str):
    """Link Identifier"""

    def __new__(cls, interface_a: InterfaceID, interface_b: InterfaceID):
        ifaces = sorted((interface_a, interface_b))
        raw_str = ":".join(ifaces)
        digest = hashlib.sha256(raw_str.encode('utf-8')).hexdigest()
        obj = super().__new__(cls, digest)
        obj.interfaces = tuple(ifaces)
        return obj

    def __init__(self, interface_a: InterfaceID, interface_b: InterfaceID):
        pass

    def __getnewargs__(self):
        """To make sure it's pickleable"""
        return self.interfaces

@dataclass
class Link:
    controller: 'Controller'
    id: InterfaceID

    @property
    def endpoint_a(self):
        iface_id: InterfaceID = self.id.interfaces[0]
        # Could also use a specialized dict for interfaces to remove a level of indirection
        return self.controller.interfaces[iface_id.switch][iface_id.port]

    @property
    def endpoint_b(self):
        iface_id: InterfaceID = self.id.interfaces[0]
        # Could also use a specialized dict for interfaces to remove a level of indirection
        return self.controller.interfaces[iface_id.switch][iface_id.port]

    @property
    def endpoints(self):
        # Would make this the main way to access instead if was using the specialized dict
        return [self.endpoint_a, self.endpoint_b]

@dataclass
class Controller:
    switches: dict[SwitchID, Switch]
    interfaces: dict[SwitchID, dict[int, Interface]]
    links: dict[LinkID, Link]
```

**Comment by viniarck** [comment]
> I've been considering something related to this for a while. What could be done is to decouple the link from its endpoints, and instead have a reference to an object representing the overall topology. Accessing the endpoints can then be done by looking at LinkID, then accessing the interfaces through the topology. LinkID just needs the interfaces sorted ahead of the time to enforce determinism, which can be done with a few simple changes.
> 
> What I am proposing should approximately look like this:
> 
> ```python
> from typing_extensions import NewType
> from dataclasses import dataclass
> import hashlib
> 
> SwitchID = NewType('SwitchID', str)
> 
> @dataclass
> class Switch:
>     controller: 'Controller'
>     id: SwitchID
> 
>     def interfaces(self):
>         return self.controller.interfaces[self.id]
> 
> class InterfaceID(str):
>     """Interface Identifier"""
> 
>     __slots__ = ("switch", "port")
> 
>     def __new__(cls, switch: SwitchID, port: int):
>         return super().__new__(cls, f"{switch}:{port}")
> 
>     def __init__(self, switch: SwitchID, port: int):
>         # Used for sorting, but can be accessed
>         self.switch = switch
>         self.port = port
>         self.tuple = (switch, port)
>         super().__init__()
> 
>     def __lt__(self, other):
>         # Ensures that instances are sortable in a way that maintains backwards
>         # compatibility when creating instances of LinkID
>         if isinstance(other, InterfaceID):
>             return self.tuple < other.tuple
>         return NotImplemented
> 
>     def __getnewargs__(self):
>         """To make sure it's pickleable"""
>         return self.tuple
> 
> @dataclass
> class Interface:
>     controller: 'Controller'
>     id: InterfaceID
> 
>     @property
>     def switch(self):
>         return self.controller.switches[self.id.switch]
> 
>     @property
>     def link(self):
>         # Yes this is very inefficent, but the alternative is to write
>         # a really complicated specialized dict type.
>         links = [link for link in self.controller.links if self.id in link.interfaces]
>         if links:
>             return self.controller.links[links[0]]
>         return None
> 
>     @property
>     def endpoint(self):
>         link = self.link
>         if link is None:
>             return None
>         endpoints = link.interfaces
>         if link.endpoint_a.id == self.id:
>             return link.endpoint_b
>         return link.endpoint_a
> 
> 
> class LinkID(str):
>     """Link Identifier"""
> 
>     def __new__(cls, interface_a: InterfaceID, interface_b: InterfaceID):
>         ifaces = sorted((interface_a, interface_b))
>         raw_str = ":".join(ifaces)
>         digest = hashlib.sha256(raw_str.encode('utf-8')).hexdigest()
>         obj = super().__new__(cls, digest)
>         obj.interfaces = tuple(ifaces)
>         return obj
> 
>     def __init__(self, interface_a: InterfaceID, interface_b: InterfaceID):
>         pass
> 
>     def __getnewargs__(self):
>         """To make sure it's pickleable"""
>         return self.interfaces
> 
> @dataclass
> class Link:
>     controller: 'Controller'
>     id: InterfaceID
> 
>     @property
>     def endpoint_a(self):
>         iface_id: InterfaceID = self.id.interfaces[0]
>         # Could also use a specialized dict for interfaces to remove a level of indirection
>         return self.controller.interfaces[iface_id.switch][iface_id.port]
> 
>     @property
>     def endpoint_b(self):
>         iface_id: InterfaceID = self.id.interfaces[0]
>         # Could also use a specialized dict for interfaces to remove a level of indirection
>         return self.controller.interfaces[iface_id.switch][iface_id.port]
> 
>     @property
>     def endpoints(self):
>         # Would make this the main way to access instead if was using the specialized dict
>         return [self.endpoint_a, self.endpoint_b]
> 
> @dataclass
> class Controller:
>     switches: dict[SwitchID, Switch]
>     interfaces: dict[SwitchID, dict[int, Interface]]
>     links: dict[LinkID, Link]
> ```

@Ktmi thanks for your suggestion, and looking into it. Let's keep it as simple as possible. 

We could assign to a Link `endpoint_a` the lowest `Interface` and on `endpoint_b` the highest one, if we were to implement the `__lt__` method that should suffice without having to introduce other changes. If you see a simpler solution feel free to proceed, thanks, appreciated your help and suggestions.

**Comment by viniarck** [comment]
I'm shifting the target to `2023.2` since this hasn't landed on `2023.1`, and there's still other tasks that David is working on. In the future, he'll continue on PR #329.
